src/packbuilder.rs

```diff
@@ -0,0 +1,358 @@
+    /// referenced trees and blobs.
+    pub fn insert_tree(&mut self, id: Oid) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_tree(self.raw, &*id.raw()));
+        }
+        Ok(())
+    }
+
+    /// Insert a commit object. This will add a commit as well as the completed
+    /// referenced tree.
+    pub fn insert_commit(&mut self, id: Oid) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_commit(self.raw, &*id.raw()));
+        }
+        Ok(())
+    }
+
+    /// Insert objects as given by the walk. Those commits and all objects they
+    /// reference will be inserted into the packbuilder.
+    pub fn insert_walk(&mut self, walk: &mut Revwalk) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_walk(self.raw, walk.raw()));
+        }
+        Ok(())
+    }
+
+    /// Recursively insert an object and its referenced objects. Insert the
+    /// object as well as any object it references.
+    pub fn insert_recursive(&mut self,
+                            id: Oid,
+                            name: Option<&str>)
+                            -> Result<(), Error> {
+        let name = try!(::opt_cstr(name));
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_recur(self.raw,
+                                                        &*id.raw(),
+                                                        name));
+        }
+        Ok(())
+    }
+
+    /// Write the contents of the packfile to an in-memory buffer. The contents
+    /// of the buffer will become a valid packfile, even though there will be
+    /// no attached index.
+    ///
+    /// `progress` will be called with progress information during pack
+    /// building. Be aware that this is called inline with pack building
+    /// operations, so performance may be affected.
+    pub fn write_buf(&mut self,
+                     buf: &mut Buf,
+                     progress: Option<&mut ProgressCb>)
+                     -> Result<(), Error> {
+        let has_progress = progress.is_some();
+        if let Some(mut progress) = progress {
+            let ptr = &mut progress as *mut _;
+            let progress_c = Some(progress_c as raw::git_packbuilder_progress);
+            unsafe {
+                try_call!(raw::git_packbuilder_set_callbacks(self.raw,
+                                                             progress_c,
+                                                             ptr as *mut _));
+            }
+        }
+        unsafe {
+            try_call!(raw::git_packbuilder_write_buf(buf.raw(), self.raw));
+        }
+        if has_progress {
+            unsafe {
+                try_call!(raw::git_packbuilder_set_callbacks(self.raw,
+                                                             None,
+                                                             ptr::null_mut()));
+            }
+        }
```

Comment:
It's not a hard requirement for my use case, but I would much prefer to have it if possible. I just thought of something that may solve both issues; by binding the lifetime `'a` into a returned guard that removes the callback when dropped I think it should be safe, it could also allow boxing the callback which might be needed to fix the double reference type inference problem. I have to prototype it and think about all the ways it could fail, especially if the guard is forgotten instead of dropped, but I hope for a signature something like:

```rust
pub fn<'a, 'b: 'a, F>(&'a mut self, cb: &'b F) -> CallbackGuard<'a> where F: Callback<'b>
```